smack-core/src/main/java/org/jivesoftware/smack/remoteEndpoint/HttpLookUp.java

```diff
@@ -0,0 +1,197 @@
+/**
+ *
+ * Copyright 2020 Aditya Borikar.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smack.remoteEndpoint;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.InetAddress;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+import org.jivesoftware.smack.datatypes.UInt16;
+import org.jivesoftware.smack.util.PacketParserUtils;
+import org.jivesoftware.smack.util.rce.RemoteConnectionEndpoint;
+import org.jivesoftware.smack.xml.XmlPullParser;
+import org.jivesoftware.smack.xml.XmlPullParserException;
+
+import org.minidns.util.InetAddressUtil;
+
+/**
+ * The HTTP lookup method uses web host metadata to list the URIs of alternative connection methods.<br>
+ * <br>
+ * To obtain endpoints for Bosh or Websocket connections, use {@link RelAttribute#BOSH} and <br>
+ * {@link RelAttribute#WEBSOCKET} respectively with {@link #getEndpoints(RelAttribute)}. <br>
+ * <br>
+ * Example:<br>
+ * <pre>
+ * {@code
+ * HttpLookUp lookup = new HttpLookUp(host,usingHttps);
+ * List<RemoteConnectionEndpoint> endpoints = lookup.getEndpoints(RelAttribute.WEBSOCKET);
+ * }
+ * </pre>
+ * @see <a href="https://xmpp.org/extensions/xep-0156.html#http">
+*     HTTP Lookup Method.
+*     </a>
+ */
+public final class HttpLookUp {
```

Comment:
This is a xep156 implementation. As such, it should go into the `org.jivesoftware.smack.altconnections` package. The class name should be `HttpLookupMethod`. I'd design it with the following static methods:

```
InputStream getXrdStream(DomainBareJid xmppServiceAddress)
List<URI> parseXrdLinkReferencesFor(XmlPullParser xmlPullParser, String relation)
List<URI> lookup(DomainBareJid xmppServiceAddress, String relation)
List<URI> lookup(DomainBareJid xmppServiceAddress, LinkRelation relation)
```

where `LinkRelation` is basically your existing `RelAttribute`.

There is no point in using a constructor (yet, that may change). Nor should this API be based on RemoteConnectionEndpoints IMHO. That is to be done by the users of this API.

Reply:
And, of course, there should be a unit test for `parseXrdLinkReferencesFor()`. The other methods are not easily unit testable, as they perform network I/O.